### PR TITLE
Add LeakyRelu C++ Op and its gradient implementation.

### PR DESCRIPTION
**Pull request by lowintelligence**
LeakyRelu, defined as 'y = { x (x>=0) or alpha*x (x<0) }', was computed by combined Ops 'max(x, alpha*x)' in current codes. Hence its gradient calculation for back propagation would contain a serial of element-wise Ops. This looks really unnecessary for such a simple op and it could be done within just one Op with less memory accesses.

In fact, we meet some performance issue on CPU in an attention based model. Within the FC calculation 'Leaky ReLu' was used as the activation function, so that its back prop Ops took some CPU time. Thus we decided to implement leaky_relu and its gradient as C++ Ops.

### CODE REVIEW

File: tensorflow/core/kernels/relu_op.h [hunk 1]
```diff
@@ -131,6 +131,65 @@ void Relu6GradOp<Device, T>::OperateNoTemplate(OpKernelContext* context,
...
+  }
+
+  void Operate(OpKernelContext* context, const Tensor& input, Tensor* output) {
+    functor::LeakyRelu<Device, T> functor;
+    functor(context->eigen_device<Device>(), input.flat<T>(),
+            alpha_, output->flat<T>());
+  }
+
+ private:
+  T alpha_;
```
Comment (akshaym) [inline tensorflow/core/kernels/relu_op.h]: Per nn_ops.cc, alpha_ is always a float right?

File: tensorflow/core/ops/ops.pbtxt [hunk 2]
```diff
@@ -13604,6 +13604,74 @@ op {
     minimum: 1
   }
 }
+op {
```
Comment (akshaym) [inline tensorflow/core/ops/ops.pbtxt]: I believe you will need to update the api_def files here: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/update_api_def.sh

File: tensorflow/python/eager/pywrap_tfe_src.cc [hunk 3]
```diff
@@ -1799,6 +1800,7 @@ bool OpDoesntRequireInput(const string& op_name) {
           "BiasAdd",
           "Relu",
           "Relu6",
+          "LeakyRelu",
```
Comment (akshaym) [inline tensorflow/python/eager/pywrap_tfe_src.cc]: The LeakyRelu gradient does seem to use the inputs? Same above.

File: tensorflow/core/ops/ops.pbtxt [hunk 4]
```diff
@@ -13604,6 +13604,74 @@ op {
...
+      list {
+        type: DT_HALF
+        type: DT_FLOAT
+        type: DT_DOUBLE
+      }
+    }
+  }
+}
+op {
+  name: "LeakykReluGrad"
```
Comment (akshaym) [inline tensorflow/core/ops/ops.pbtxt]: I don't believe these changes are required (they are auto generated), but in case they are, you have a typo here - LeakykReluGrad -> LeakyReluGrad.

**Review by akshaym — COMMENTED** [review]
Sorry for the delay @lowintelligence, and thanks for the contribution. 

I've just added some initial comments, and requested a more in-depth review.

File: tensorflow/core/kernels/relu_op.h (hunk 1 above)
Comment (lowintelligence) [reply tensorflow/core/kernels/relu_op.h]: Well, here I checked 'tensorflow/core/kernels/lrn_op.cc' as a reference and kept the same style with that. Since LeakyRelu C++ Op would only support {half, float, double} type, thus convert alpha type from float to T should be reasonable.

File: tensorflow/core/ops/ops.pbtxt (hunk 2 above)
Comment (lowintelligence) [reply tensorflow/core/ops/ops.pbtxt]: Sorry, I wrote the api_def files manually but forgot to 'git add' them. I just re-generated the files with the tool and add them to the new commit.

File: tensorflow/python/eager/pywrap_tfe_src.cc (hunk 3 above)
Comment (lowintelligence) [reply tensorflow/python/eager/pywrap_tfe_src.cc]: My fault, 'LeakyRelu' and 'LeakyReluGrad' both use the input and don't use the output, corrected in the new commit.

File: tensorflow/core/ops/ops.pbtxt (hunk 4 above)
Comment (lowintelligence) [reply tensorflow/core/ops/ops.pbtxt]: Really a typo. Thanks for figuring it out. :)

File: tensorflow/python/ops/nn_ops.py [hunk 5]
```diff
@@ -1601,8 +1601,7 @@ def leaky_relu(features, alpha=0.2, name=None):
     features = ops.convert_to_tensor(features, name="features")
     if features.dtype.is_integer:
       features = math_ops.to_float(features)
-    alpha = ops.convert_to_tensor(alpha, dtype=features.dtype, name="alpha")
```
Comment (alextp) [inline tensorflow/python/ops/nn_ops.py]: This should be changed to respect our forward compatibility rules. Use the forward compatibility layer to make this change, with the date set 3 weeks in the future of the merge (so a month from now is safe): https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/compat/compat.py

File: tensorflow/python/ops/nn_grad.py [hunk 6]
```diff
@@ -390,6 +390,21 @@ def _Relu6GradGrad(op, grad):
...
 
+@ops.RegisterGradient("LeakyRelu")
+def _LeakyReluGrad(op, grad):
+  x = op.inputs[0]
+  alpha = op.get_attr("alpha")
+  return gen_nn_ops.leaky_relu_grad(grad, x, alpha=alpha)
+
+
+@ops.RegisterGradient("LeakyReluGrad")
+def _LeakyReluGradGrad(op, grad):
```
Comment (alextp) [inline tensorflow/python/ops/nn_grad.py]: Can you add a gradient checker test for these? Use https://www.tensorflow.org/api_docs/python/tf/test/compute_gradient_error

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/cc/gradients/nn_grad.cc [hunk 7]
```diff
@@ -143,6 +143,19 @@ Status Relu6GradHelper(const Scope& scope, const Operation& op,
...
 REGISTER_GRADIENT_OP("Relu6", Relu6GradHelper);
 
+Status LeakyReluGradHelper(const Scope& scope, const Operation& op,
+                           const std::vector<Output>& grad_inputs,
+                           std::vector<Output>* grad_outputs) {
+  float alpha;
+  TF_RETURN_IF_ERROR(GetNodeAttr(op.node()->attrs(), "alpha", &alpha));
+  internal::LeakyReluGrad::Attrs attrs;
+  attrs.Alpha(alpha);
+  auto dx = internal::LeakyReluGrad(scope, grad_inputs[0], op.input(0), attrs);
```
[the remark this answers is not in the document]
Comment (lowintelligence) [reply tensorflow/cc/gradients/nn_grad.cc]: Done in new commit.

File: tensorflow/python/ops/nn_ops.py (hunk 5 above)
Comment (lowintelligence) [reply tensorflow/python/ops/nn_ops.py]: Well, I tried to use compat.forword_compatible() on my new implementation. However, another concern (maybe over considered) is that, whether somebody would use 'alpha' as a trainable variable in his/her model? Should we try to keep compatibility for this case?

File: tensorflow/python/ops/nn_grad.py (hunk 6 above)
Comment (lowintelligence) [reply tensorflow/python/ops/nn_grad.py]: Well, I believe that python UTs for GradGrad was already in the original commit.

**Review by alextp — DISMISSED** [review]

**Comment by lowintelligence** [comment]
Due to ci_build error reports, I did some change and submit a new version here.

File: tensorflow/core/api_def/base_api/api_def_LeakyRelu.pbtxt [hunk 8]
```diff
@@ -0,0 +1,4 @@
+op {
+  graph_op_name: "LeakyRelu"
+  summary: "Computes rectified linear: `max(features, features * alpha)`."
+}
```
Comment (alextp) [inline tensorflow/core/api_def/base_api/api_def_LeakyRelu.pbtxt]: Can you add visibility:hidden here?

I think this will obviate the need for the API golden file changes since there's already tf.nn.leaky_relu and no need for tf.leaky_relu (which gets implicitly created when an op is added)

**Comment by lowintelligence** [comment]
@alextp Do you have time to review the code again? Thanks.

**Review by alextp — DISMISSED** [review]

**Comment by lowintelligence** [comment]
@alextp Sorry, I have to add a commit to fix some lint errors and typos due to ci_build tests report.

**Comment by lowintelligence** [comment]
Have to merge with master due to branch conflicts.

**Review by alextp — DISMISSED** [review]

**Comment by lowintelligence** [comment]
Have to reverted XLA commit since the Op didn't provide bfloat16 support.

**Comment by lowintelligence** [comment]
@alextp I reverted the XLA commit due to unsupported dtypes. Finally it looks that all CI build cases related with this PR was passed. Thanks for the patient of taking a review again. :)

**Comment by facaiy** [comment]
Indeed, unrelated failures. Thanks for your work, @lowintelligence !

**Review by alextp — DISMISSED** [review]

**Comment by lowintelligence** [comment]
@alextp Could you please help to take a look about why "feedback/copybara" failed? I can't see detail information here.

**Comment by facaiy** [comment]
@yifeif Hi, yifei. Do we have a plan to give contributor more details about feedback/copybara? It seems that we can't do anything but wait. cc @martinwicke

**Comment by yifeif** [comment]
Hi @facaiy feedback/copybara is currently an experimental feature and it's a bit far from being able to show the build results with details. Please ping reviewer(s) to see if there is anything needs to be fixed from the PR. Thanks!

**Comment by facaiy** [comment]
@yifeif Oh, I got it.  It can be quite useful when feddback/copybara is ready for showing build details. Thanks for your explanation :-)

**Comment by facaiy** [comment]
@protoget Hi, could you lend a hand? Thanks.

**Comment by alextp** [comment]
One internal error is error: module //tensorflow/cc:gradients_nn_grad_test does not depend on a module exporting 'tensorflow/cc/ops/nn_ops_internal.h'

I think you need to fix the build dependencies, adding something to tensorflow/cc/BUILD

**Comment by lowintelligence** [comment]
Well, I've fixed the tensorflow/cc/BUILD file and delayed the forward compatibility date. Please take a check. Thanks. @alextp

**Review by alextp — DISMISSED** [review]

**Comment by lowintelligence** [comment]
Well, I found that the commit in master branch https://github.com/tensorflow/tensorflow/commit/e2ce9787d9927e4a6574e6ac4606a47712320170 has removed the 'compat' module importing in nn_ops.py, thus many tests were failed due to this module not found. :-( I would make a merge with master and commit a new version. @alextp Thanks.

**Review by alextp — APPROVED** [review]
